src/tools/rbd/action/Export.cc

```diff
@@ -195,7 +195,7 @@ int do_export_diff_fd(librbd::Image& image, const char *fromsnapname,
   ExportDiffContext edc(&image, fd, info.size,
                         g_conf().get_val<uint64_t>("rbd_concurrent_management_ops"),
                         no_progress, export_format);
-  r = image.diff_iterate2(fromsnapname, 0, info.size, true, whole_object,
+  r = image.diff_iterate2(fromsnapname, 0, info.size, include_parent, whole_object,
```

Comment:
I don't think this would work for whiteout objects (i.e. 0-size objects in the clone that are used to indicate that a corresponding range in the clone has been written to) because the diff with `include_parent=false` doesn't make a distinction between a nonexistent object and 0-size object.  Consider the following:
```
$ rbd clone parentimg@snap cloneimg
$ rbd export --exclude-parent cloneimg ... 
... GOOD: export file is empty, after importing cloneimg-imported matches cloneimg
$ sudo rbd device map cloneimg
/dev/rbd0
... write some data to /dev/rbd0 ...
$ rbd export --exclude-parent cloneimg ...
... GOOD: export file has some data, after importing cloneimg-imported matches cloneimg
$ sudo blkdiscard /dev/rbd0
$ rbd export --exclude-parent cloneimg ...
... BAD: export file is empty, after importing cloneimg-imported matches cloneimg
    reads from cloneimg return zeroes
    reads from cloneimg-imported return parent data
```

Reply:
(Disclaimer: I haven't actually compiled the PR and tried the above, so there is a chance that I'm wrong.  I just wanted to present this case for consideration.)

Reply:
Thank you, I conducted the test as you said, and the results are as described by you。

Reply:
@idryomov   Can I add a new api diff_iterate_from_parent to calculate the difference from parent ?
Only a small modification is needed in compute_diffs.

```
 if (diff.empty() && !whole_object) {
    // return parent overlap when from_parent and diff.empty().
    if (m_diff_context.from_snap_id == 0 && (!end_exists || m_diff_context.from_parent)){
      compute_parent_overlap(diffs);
    }
    return;
  }

  interval_set<uint64_t> overlap;  // object extents
  overlap.insert(opos, r->second);
  overlap.intersection_of(diff);
  
  // return union of parent overlap and diff overlap
  if (m_diff_context.from_parent && !diff_context.parent_diff.empty()) {
    interval_set<uint64_t> overlap_parent;
    overlap_parent.insert(m_offset + r->first, r->second);
    overlap_parent.intersection_of(m_diff_context.parent_diff);
    if (!overlap_parent.empty()) {
      for (interval_set<uint64_t>::iterator i = overlap_parent.begin();
           i != overlap_parent.end(); ++i) {
        overlap.insert(i.get_start() - m_offset, i.get_len());
      }
    }
  }

```

Reply:
@idryomov I’ve solved the issue of how to calculate the differences between a clone rbd and its parent.  Could you take a look and help review it? https://github.com/ceph/ceph/pull/60449